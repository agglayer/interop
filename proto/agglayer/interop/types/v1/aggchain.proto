syntax = "proto3";

package agglayer.interop.types.v1;

import "agglayer/interop/types/v1/bytes.proto";

// Aggchaindata is the data sent by the aggchain.
message AggchainData {
  // The data of the aggchain.
  oneof data {
    // Legacy ECDSA signature on a commitment to the state transition.
    FixedBytes65 signature = 1;

    // Aggchain proof with the single ECDSA verified by the agglayer only.
    AggchainProof generic = 2;

    // Only multisig performed by the chain's committee.
    Multisig multisig = 3;

    // Multisig combined with an aggchain proof.
    AggchainProofWithMultisig multisig_and_aggchain_proof = 4;
  }
}

// Multisig
message Multisig {
  // Multisig scheme.
  oneof data {
    // ECDSA multisig scheme.
    ECDSAMultisig ecdsa = 1;
  }
}

// ECDSA multisig scheme with signatures from the committee members.
message ECDSAMultisig {
  // Represents the key-value pair of a committee member signature entry.
  message ECDSAMultisigEntry {
    // The key is the place of the signer in the list of multisig signer on L1.
    uint32 index = 1;
    // The value is the signature made by the commitee member.
    FixedBytes65 signature = 2;
  }
  // Signatures from the committee members.
  repeated ECDSAMultisigEntry signatures = 3;
}

// Multisig and aggchain proof
message AggchainProofWithMultisig {
  // Multisig
  Multisig multisig = 1;

  // AggchainProof
  AggchainProof aggchain_proof = 2;
}

// Aggchain proof.
message AggchainProof {
  // Aggchain params.
  FixedBytes32 aggchain_params = 1;

  // Important: The signature is ignored if the aggchain proof is part of a AggchainProofWithMultisig.
  // The signature linked to the aggchain-proof.
  // This signature is used to verify the authenticity of the aggchain-proof.
  // It represents a commitment to:
  // - `NewLocalExitRoot`
  // - `ClaimHash` (which is the hash of all the claims's global index + bridge exit hash)
  // - `Height` of the certificate associated to this aggchain-proof
  // - `AggchainParams` of the aggchain-proof
  FixedBytes65 signature = 2;

  // Aggchain proof context to share non-sensitive data.
  map<string, bytes> context = 3;
  // The proof for the aggchain.
  oneof proof {
    // SP1 stark proof.
    SP1StarkProof sp1_stark = 4;
  }
}

// SP1 stark proof.
message SP1StarkProof {
  // SP1 Version
  string version = 1;

  // SP1 stark proof.
  bytes proof = 2;

  // SP1 stark proof verification key.
  bytes vkey = 3;
}
