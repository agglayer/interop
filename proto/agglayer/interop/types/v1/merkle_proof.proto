syntax = "proto3";

package agglayer.interop.types.v1;

import "agglayer/interop/types/v1/bytes.proto";

// Represents a Merkle proof.
message MerkleProof {
  // The root of the Merkle tree.
  FixedBytes32 root = 1;
  // The path from the leaf to the root.
  repeated FixedBytes32 siblings = 2;
}
