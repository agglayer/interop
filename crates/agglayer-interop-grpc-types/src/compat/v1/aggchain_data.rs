use std::collections::HashMap;

use agglayer_interop_types::aggchain_proof::{AggchainData, Proof, SP1StarkWithContext};
use bincode::Options as _;
use prost::bytes::Bytes;

use super::Error;
use crate::v1;

#[inline]
fn sp1v4_bincode_options() -> impl bincode::Options {
    bincode::DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

impl TryFrom<v1::AggchainData> for AggchainData {
    type Error = Error;

    fn try_from(value: v1::AggchainData) -> Result<Self, Self::Error> {
        Ok(match value.data {
            Some(v1::aggchain_data::Data::Signature(signature)) => AggchainData::ECDSA {
                signature: (&*signature.value)
                    .try_into()
                    .map_err(Error::parsing_signature)?,
            },
            Some(v1::aggchain_data::Data::Generic(proof)) => AggchainData::Generic {
                aggchain_params: required_field!(proof, aggchain_params),
                proof: match proof.proof {
                    Some(v1::aggchain_proof::Proof::Sp1Stark(v1::Sp1StarkProof {
                        version,
                        proof,
                        vkey,
                    })) => Proof::SP1Stark(SP1StarkWithContext {
                        proof: Box::new(
                            std::panic::catch_unwind(|| {
                                sp1v4_bincode_options().deserialize(&proof)
                            })
                            .map_err(|_| {
                                Error::deserializing_proof(Box::new(bincode::ErrorKind::Custom(
                                    String::from("panic"),
                                )))
                            })?
                            .map_err(Error::deserializing_proof)?,
                        ),
                        vkey: std::panic::catch_unwind(|| {
                            sp1v4_bincode_options().deserialize(&vkey)
                        })
                        .map_err(|_| {
                            Error::deserializing_proof(Box::new(bincode::ErrorKind::Custom(
                                String::from("panic"),
                            )))
                        })?
                        .map_err(Error::deserializing_vkey)?,
                        version,
                    }),
                    None => return Err(Error::missing_field("proof").inside_field("data")),
                },
            },
            None => return Err(Error::missing_field("data")),
        })
    }
}

impl TryFrom<AggchainData> for v1::AggchainData {
    type Error = Error;

    fn try_from(value: AggchainData) -> Result<Self, Self::Error> {
        Ok(v1::AggchainData {
            data: Some(match value {
                AggchainData::ECDSA { signature } => {
                    v1::aggchain_data::Data::Signature(v1::FixedBytes65 {
                        value: Bytes::copy_from_slice(&signature.as_bytes()),
                    })
                }
                AggchainData::Generic {
                    proof: Proof::SP1Stark(proof),
                    aggchain_params,
                } => v1::aggchain_data::Data::Generic(v1::AggchainProof {
                    context: HashMap::new(),
                    aggchain_params: Some(aggchain_params.into()),
                    proof: Some(v1::aggchain_proof::Proof::Sp1Stark(v1::Sp1StarkProof {
                        version: proof.version,
                        proof: sp1v4_bincode_options()
                            .serialize(&proof.proof)
                            .map_err(Error::serializing_proof)?
                            .into(),
                        vkey: sp1v4_bincode_options()
                            .serialize(&proof.vkey)
                            .map_err(Error::serializing_vkey)?
                            .into(),
                    })),
                }),
            }),
        })
    }
}
