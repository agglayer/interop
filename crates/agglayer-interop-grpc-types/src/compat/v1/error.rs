use std::fmt;

use agglayer_interop_types::primitives::SignatureError;
use tonic_types::FieldViolation;

#[derive(Debug, thiserror::Error)]
pub enum SourceError {
    #[error(transparent)]
    Bincode(#[from] bincode::Error),

    #[error(transparent)]
    Signature(#[from] SignatureError),
}

#[derive(Clone, Copy, Debug)]
pub enum ErrorKind {
    InvalidData,
    MissingField,
}

impl fmt::Display for ErrorKind {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ErrorKind::InvalidData => {
                write!(f, "Invalid data")
            }
            ErrorKind::MissingField => {
                write!(f, "Missing field")
            }
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub struct Error {
    kind: ErrorKind,
    message: String,
    field: Vec<&'static str>,
    #[source]
    source: Option<SourceError>,
}

impl fmt::Display for Error {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if !self.field.is_empty() {
            write!(f, "{}: ", self.field_str())?;
        }
        write!(f, "{}", self.message)
    }
}

impl Error {
    #[inline]
    pub fn missing_field(f: &'static str) -> Self {
        Error {
            kind: ErrorKind::MissingField,
            message: "required field is missing".to_string(),
            field: vec![f],
            source: None,
        }
    }

    #[inline]
    pub fn invalid_data(m: String) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: m,
            field: vec![],
            source: None,
        }
    }

    #[inline]
    pub fn inside_field(mut self, f: &'static str) -> Self {
        self.field.push(f);
        self.field.rotate_right(1);
        self
    }

    #[inline]
    pub fn serializing_proof(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to serialize proof".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn serializing_vkey(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to serialize vkey".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn serializing_context(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to serialize context".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn deserializing_proof(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to deserialize proof".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn deserializing_vkey(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to deserialize vkey".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn deserializing_aggchain_proof_public_values(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to deserialize aggchain proof public values".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn serializing_aggchain_proof_public_values(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to serialize aggchain proof public values".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    #[inline]
    pub fn parsing_signature(e: SignatureError) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to parse signature".to_string(),
            field: vec![],
            source: Some(SourceError::Signature(e)),
        }
    }

    #[inline]
    pub fn kind(&self) -> ErrorKind {
        self.kind
    }

    #[inline]
    pub fn field(&self) -> &[&'static str] {
        &self.field
    }

    #[inline]
    pub fn field_str(&self) -> String {
        if self.field.is_empty() {
            ".".to_string()
        } else {
            self.field.join(".")
        }
    }
}

impl From<&Error> for Vec<FieldViolation> {
    #[inline]
    fn from(value: &Error) -> Self {
        vec![FieldViolation::new(
            value.field_str(),
            value.message.clone(),
        )]
    }
}
