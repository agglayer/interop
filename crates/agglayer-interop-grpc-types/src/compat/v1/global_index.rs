use agglayer_interop_types::{GlobalIndex, U256};

use super::Error;
use crate::v1;

impl TryFrom<v1::FixedBytes32> for GlobalIndex {
    type Error = Error;

    #[inline]
    fn try_from(value: v1::FixedBytes32) -> Result<Self, Self::Error> {
        Ok(GlobalIndex::from(U256::try_from(value)?))
    }
}

impl From<GlobalIndex> for v1::FixedBytes32 {
    #[inline]
    fn from(value: GlobalIndex) -> Self {
        <U256 as From<GlobalIndex>>::from(value).into()
    }
}
