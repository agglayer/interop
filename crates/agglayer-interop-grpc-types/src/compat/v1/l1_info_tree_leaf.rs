use agglayer_interop_types::L1InfoTreeLeaf;

use super::Error;
use crate::v1;

impl TryFrom<v1::L1InfoTreeLeafWithContext> for L1InfoTreeLeaf {
    type Error = Error;

    #[inline]
    fn try_from(value: v1::L1InfoTreeLeafWithContext) -> Result<Self, Self::Error> {
        Ok(L1InfoTreeLeaf {
            l1_info_tree_index: value.l1_info_tree_index,
            rer: required_field!(value, rer),
            mer: required_field!(value, mer),
            block_hash: required_field!(value, block_hash),
            timestamp: value.timestamp,
        })
    }
}

impl From<L1InfoTreeLeaf> for v1::L1InfoTreeLeafWithContext {
    #[inline]
    fn from(value: L1InfoTreeLeaf) -> Self {
        v1::L1InfoTreeLeafWithContext {
            l1_info_tree_index: value.l1_info_tree_index,
            rer: Some(value.rer.into()),
            mer: Some(value.mer.into()),
            block_hash: Some(value.block_hash.into()),
            timestamp: value.timestamp,
        }
    }
}
