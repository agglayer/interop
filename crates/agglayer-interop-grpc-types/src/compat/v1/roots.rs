use agglayer_interop_types::roots::LocalExitRoot;
use prost::bytes::Bytes;

use super::Error;
use crate::v1;

impl TryFrom<v1::FixedBytes32> for LocalExitRoot {
    type Error = Error;

    #[inline]
    fn try_from(value: v1::FixedBytes32) -> Result<Self, Self::Error> {
        Ok(LocalExitRoot::from(<[u8; 32]>::try_from(value)?))
    }
}

impl From<LocalExitRoot> for v1::FixedBytes32 {
    #[inline]
    fn from(value: LocalExitRoot) -> Self {
        v1::FixedBytes32 {
            value: Bytes::copy_from_slice(value.as_ref()),
        }
    }
}
