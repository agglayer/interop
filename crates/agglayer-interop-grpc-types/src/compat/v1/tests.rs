use agglayer_interop_types::{
    aggchain_proof::AggchainData, primitives::SignatureError, Address, BridgeExit,
    ClaimFromMainnet, ClaimFromRollup, Digest, GlobalIndex, ImportedBridgeExit, L1InfoTreeLeaf,
    L1InfoTreeLeafInner, MerkleProof, TokenInfo, U256,
};
use prost::Message;

use super::Error;
use crate::v1;

#[rstest::rstest]
#[case::error("no_proof", Error::missing_field("proof"))]
#[case::error("bad_data", Error::invalid_data("invalid value".to_owned()))]
#[case::error("bad_data_in_field", Error::invalid_data("invalid value".to_owned()).inside_field("value"))]
#[case::error("bad_data_in_nested", Error::invalid_data("invalid value".to_owned()).inside_field("value").inside_field("data"))]
#[case::error("failed_ser", Error::serializing_proof(Box::new(bincode::ErrorKind::Custom("failed".to_owned()))))]
#[case::error("failed_deser", Error::deserializing_proof(Box::new(bincode::ErrorKind::Custom("failed".to_owned()))))]
#[case::error("bad_sig", Error::parsing_signature(SignatureError::InvalidParity(5)))]
#[case::error("bad_sig_in_nested", Error::parsing_signature(SignatureError::InvalidParity(5)).inside_field("signature").inside_field("data"))]
fn error_messages(#[case] name: &str, #[case] error: Error) {
    insta::assert_snapshot!(format!("{name}/display"), error);
    insta::assert_debug_snapshot!(format!("{name}/kind"), error.kind());
    insta::assert_snapshot!(
        format!("{name}/debug"),
        format!("{:?}", eyre::Error::from(error))
    );
}

macro_rules! make_parser_fuzzers {
    ($test:ident, $proto:ty, $type:ty) => {
        #[test]
        fn $test() {
            bolero::check!().for_each(|bytes| {
                if let Ok(proto) = <$proto>::decode(bytes) {
                    let _ = <$type>::try_from(proto);
                };
            })
        }
    };
}

make_parser_fuzzers!(fuzz_parser_address, v1::FixedBytes20, Address);
make_parser_fuzzers!(fuzz_parser_aggchain_data, v1::AggchainData, AggchainData);
make_parser_fuzzers!(fuzz_parser_bridge_exit, v1::BridgeExit, BridgeExit);
make_parser_fuzzers!(
    fuzz_parser_claim_from_mainnet,
    v1::ClaimFromMainnet,
    ClaimFromMainnet
);
make_parser_fuzzers!(
    fuzz_parser_claim_from_rollup,
    v1::ClaimFromRollup,
    ClaimFromRollup
);
make_parser_fuzzers!(fuzz_parser_digest, v1::FixedBytes32, Digest);
make_parser_fuzzers!(fuzz_parser_global_index, v1::FixedBytes32, GlobalIndex);
make_parser_fuzzers!(
    fuzz_parser_imported_bridge_exit,
    v1::ImportedBridgeExit,
    ImportedBridgeExit
);
make_parser_fuzzers!(
    fuzz_parser_l1_info_tree_leaf_with_context,
    v1::L1InfoTreeLeafWithContext,
    L1InfoTreeLeaf
);
make_parser_fuzzers!(
    fuzz_parser_l1_info_tree_leaf_inner,
    v1::L1InfoTreeLeaf,
    L1InfoTreeLeafInner
);
make_parser_fuzzers!(fuzz_parser_merkle_proof, v1::MerkleProof, MerkleProof);
make_parser_fuzzers!(fuzz_parser_token_info, v1::TokenInfo, TokenInfo);
make_parser_fuzzers!(fuzz_parser_u256, v1::FixedBytes32, U256);

macro_rules! make_round_trip_fuzzers {
    ($test:ident, $proto:ty, $type:ty) => {
        #[test]
        fn $test() {
            bolero::check!()
                .with_arbitrary::<$type>()
                .for_each(|input: &$type| {
                    let proto: $proto = input.clone().into();
                    let output = <$type>::try_from(proto).unwrap();
                    assert_eq!(input, &output);
                })
        }
    };
}

make_round_trip_fuzzers!(fuzz_round_trip_address, v1::FixedBytes20, Address);
#[test]
fn fuzz_round_trip_aggchain_data() {
    bolero::check!()
        .with_arbitrary::<AggchainData>()
        .for_each(|input| {
            let proto: v1::AggchainData = input.clone().try_into().unwrap();

            // Check if input has empty multisig signatures
            let has_empty_multisig = match &input {
                AggchainData::MultisigOnly { multisig } => multisig.0.is_empty(),
                AggchainData::MultisigAndAggchainProof { multisig, .. } => multisig.0.is_empty(),
                _ => false,
            };

            match AggchainData::try_from(proto) {
                Ok(output) => {
                    // If we had empty multisig, this should not have succeeded
                    assert!(
                        !has_empty_multisig,
                        "Expected error for empty multisig signatures, but conversion succeeded"
                    );

                    // For most cases, we can't check equality due to stark proofs,
                    // but we can at least verify the conversion succeeded
                    if let (
                        AggchainData::ECDSA { signature: sig1 },
                        AggchainData::ECDSA { signature: sig2 },
                    ) = (&input, &output)
                    {
                        assert_eq!(sig1, sig2);
                    }
                }
                Err(err) => {
                    if has_empty_multisig {
                        // This error is expected for empty multisig cases
                        let err_msg = err.to_string();
                        assert!(
                            err_msg.contains("Multisig ECDSA doesn't have any signature"),
                            "Expected empty multisig error, got: {}",
                            err
                        );
                    } else {
                        // Any other error should cause the test to fail
                        panic!("Unexpected conversion error: {}", err);
                    }
                }
            }
        })
}
make_round_trip_fuzzers!(fuzz_round_trip_bridge_exit, v1::BridgeExit, BridgeExit);
make_round_trip_fuzzers!(
    fuzz_round_trip_claim_from_mainnet,
    v1::ClaimFromMainnet,
    ClaimFromMainnet
);
make_round_trip_fuzzers!(
    fuzz_round_trip_claim_from_rollup,
    v1::ClaimFromRollup,
    ClaimFromRollup
);
make_round_trip_fuzzers!(fuzz_round_trip_digest, v1::FixedBytes32, Digest);
make_round_trip_fuzzers!(fuzz_round_trip_global_index, v1::FixedBytes32, GlobalIndex);
make_round_trip_fuzzers!(
    fuzz_round_trip_imported_bridge_exit,
    v1::ImportedBridgeExit,
    ImportedBridgeExit
);
make_round_trip_fuzzers!(
    fuzz_round_trip_l1_info_tree_leaf_with_context,
    v1::L1InfoTreeLeafWithContext,
    L1InfoTreeLeaf
);
make_round_trip_fuzzers!(
    fuzz_round_trip_l1_info_tree_leaf_inner,
    v1::L1InfoTreeLeaf,
    L1InfoTreeLeafInner
);
make_round_trip_fuzzers!(fuzz_round_trip_merkle_proof, v1::MerkleProof, MerkleProof);
make_round_trip_fuzzers!(fuzz_round_trip_token_info, v1::TokenInfo, TokenInfo);
make_round_trip_fuzzers!(fuzz_round_trip_u256, v1::FixedBytes32, U256);
