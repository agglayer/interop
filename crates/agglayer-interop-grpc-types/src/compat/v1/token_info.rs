use agglayer_interop_types::TokenInfo;

use super::Error;
use crate::v1;

impl TryFrom<v1::TokenInfo> for TokenInfo {
    type Error = Error;

    #[inline]
    fn try_from(value: v1::TokenInfo) -> Result<Self, Self::Error> {
        Ok(TokenInfo {
            origin_network: value.origin_network.into(),
            origin_token_address: required_field!(value, origin_token_address),
        })
    }
}

impl From<TokenInfo> for v1::TokenInfo {
    #[inline]
    fn from(value: TokenInfo) -> Self {
        v1::TokenInfo {
            origin_network: value.origin_network.to_u32(),
            origin_token_address: Some(value.origin_token_address.into()),
        }
    }
}
