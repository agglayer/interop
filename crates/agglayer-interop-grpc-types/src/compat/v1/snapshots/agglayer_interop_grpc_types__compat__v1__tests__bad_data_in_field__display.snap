---
source: crates/agglayer-interop-grpc-types/src/compat/v1/tests.rs
expression: error
---
value: invalid value
